src/geth.rs

```diff
@@ -0,0 +1,197 @@
+};
+use ethereum_types::U256;
+use num::{BigUint, Num};
+use serde_hex::{CompactPfx, SerHex};
+
+#[derive(Debug, Clone)]
+pub struct Client {
+    rpc_client: jsonrpc::Client,
+    jsonrpc_version: String,
+}
+
+impl Client {
+    pub fn new(url: reqwest::Url) -> Self {
+        Client {
+            rpc_client: jsonrpc::Client::new(url),
+            jsonrpc_version: String::from("2.0"),
+        }
+    }
+
+    pub async fn chain_id(&self) -> anyhow::Result<ChainId> {
+        let chain_id = self
+            .rpc_client
+            .send::<Vec<()>, String>(jsonrpc::Request::new(
+                "net_version",
+                vec![],
+                self.jsonrpc_version.clone(),
+            ))
+            .await
+            .context("failed to fetch net version")?;
+        let chain_id: u32 = chain_id.parse()?;
+        let chain_id = ChainId::from(chain_id);
+
+        Ok(chain_id)
+    }
+
+    pub async fn send_transaction(&self, request: SendTransactionRequest) -> anyhow::Result<Hash> {
+        let tx_hash = self
+            .rpc_client
+            .send(jsonrpc::Request::new(
+                "eth_sendTransaction",
+                vec![jsonrpc::serialize(request)?],
+                self.jsonrpc_version.clone(),
+            ))
+            .await
+            .context("failed to send transaction")?;
+
+        Ok(tx_hash)
+    }
+
+    pub async fn send_raw_transaction(&self, transaction_hex: String) -> anyhow::Result<Hash> {
+        let tx_hash = self
+            .rpc_client
+            .send(jsonrpc::Request::new(
+                "eth_sendRawTransaction",
+                vec![transaction_hex],
+                self.jsonrpc_version.clone(),
+            ))
+            .await
+            .context("failed to send raw transaction")?;
+
+        Ok(tx_hash)
+    }
+
+    pub async fn get_transaction_by_hash(
+        &self,
+        transaction_hash: Hash,
+    ) -> anyhow::Result<Transaction> {
+        let transaction = self
+            .rpc_client
+            .send(jsonrpc::Request::new(
+                "eth_getTransactionByHash",
+                vec![jsonrpc::serialize(transaction_hash)?],
+                self.jsonrpc_version.clone(),
+            ))
+            .await
+            .context("failed to get transaction by hash")?;
+
+        Ok(transaction)
+    }
+
+    pub async fn get_transaction_receipt(
+        &self,
+        transaction_hash: Hash,
+    ) -> anyhow::Result<TransactionReceipt> {
+        let receipt = self
+            .rpc_client
+            .send(jsonrpc::Request::new(
+                "eth_getTransactionReceipt",
+                vec![jsonrpc::serialize(transaction_hash)?],
+                self.jsonrpc_version.clone(),
+            ))
+            .await
+            .context("failed to get transaction receipt")?;
+
+        Ok(receipt)
+    }
+
+    pub async fn get_transaction_count(&self, account: Address) -> anyhow::Result<u32> {
+        let count: String = self
```

Comment:
[This should return a number,](https://web3js.readthedocs.io/en/v1.2.7/web3-eth.html#gettransactioncount) it does not?

Reply:
It returns a hex string :)))

Reply:
Maybe a todo here to actually cast it to a number using some hex deserialisation?

Reply:
The casting happens a few lines down:

```rust
let count = u32::from_str_radix(&count[2..], 16)?;
```

But eventually we could/should remove the deserialization logic from this function.